pkg/controller/product/application_plan_reconciler.go

```diff
@@ -0,0 +1,391 @@
+package product
+
+import (
+	"fmt"
+	"strconv"
+
+	capabilitiesv1beta1 "github.com/3scale/3scale-operator/pkg/apis/capabilities/v1beta1"
+	"github.com/3scale/3scale-operator/pkg/helper"
+	"github.com/3scale/3scale-operator/pkg/reconcilers"
+
+	threescaleapi "github.com/3scale/3scale-porta-go-client/client"
+	"github.com/go-logr/logr"
+)
+
+type limitKey struct {
+	Period   string
+	Value    int
+	MetricID int64
+}
+
+type pricingRuleKey struct {
+	From         int
+	To           int
+	PricePerUnit string
+	MetricID     int64
+}
+
+type applicationPlanReconciler struct {
+	*reconcilers.BaseReconciler
+	systemName          string
+	resource            capabilitiesv1beta1.ApplicationPlanSpec
+	productEntity       *helper.ProductEntity
+	backendRemoteIndex  *helper.BackendAPIRemoteIndex
+	planEntity          *helper.ApplicationPlanEntity
+	threescaleAPIClient *threescaleapi.ThreeScaleClient
+	logger              logr.Logger
+}
+
+func newApplicationPlanReconciler(b *reconcilers.BaseReconciler,
+	systemName string,
+	resource capabilitiesv1beta1.ApplicationPlanSpec,
+	threescaleAPIClient *threescaleapi.ThreeScaleClient,
+	productEntity *helper.ProductEntity,
+	backendRemoteIndex *helper.BackendAPIRemoteIndex,
+	planEntity *helper.ApplicationPlanEntity,
+	logger logr.Logger,
+) *applicationPlanReconciler {
+
+	return &applicationPlanReconciler{
+		BaseReconciler:      b,
+		systemName:          systemName,
+		resource:            resource,
+		threescaleAPIClient: threescaleAPIClient,
+		productEntity:       productEntity,
+		backendRemoteIndex:  backendRemoteIndex,
+		planEntity:          planEntity,
+		logger:              logger.WithValues("Plan", systemName),
+	}
+}
+
+// Reconcile ensures plan attrs, limits and pricingRules are reconciled
+func (a *applicationPlanReconciler) Reconcile() error {
+	taskRunner := helper.NewTaskRunner(nil, a.logger)
+	taskRunner.AddTask("SyncPlan", a.syncPlan)
+	taskRunner.AddTask("SyncLimits", a.syncLimits)
+	taskRunner.AddTask("SyncPricingRules", a.syncPricingRules)
+
+	err := taskRunner.Run()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *applicationPlanReconciler) syncPlan(_ interface{}) error {
+	params := threescaleapi.Params{}
+
```

Comment:
I understand we only try to add reconciliation code for fields that can be changed in the 3scale API side right?

For example, if we have a field in the CR but that field cannot be changed once the 3scale API entity is created, we have not added code to reconcile nor code verifying that the user is not doing anything "incorrect".

So, if a user changes a "create-only" field we simply ignore the change. Is that correct?

Reply:
One example is the `system-name`. It is a create only attr that 3scale API does not allow to update. The operator will not try to reconciliate, it will use as ID. If you change it in you spec, new element will be created or some other element can be reconciled.

So, yes. The reconlliation logic does take this kind of attributes into account and operate accordingly. If not, it is a bug.

Reply:
:+1: